File: src/046/p046.py
import itertools
from src.utils.py.prime import generate_primes, is_prime


def is_sum_of_prime_and_twice_a_square(
    number: int, primes: list[int], double_squares: list[int]
) -> bool:
    # Check if there is any double square that when subtracted from the number
    # yields a prime number
    filtered = itertools.takewhile(lambda dsq: dsq <= number, double_squares)
    return any(is_prime(number - dsq, primes) for dsq in filtered)


def solve() -> int:
    # Assume there is a solution below 1 million
    double_squares = [2 * i * i for i in range(1, 1000)]
    primes = generate_primes(1000)
    for number in itertools.count(start=9, step=2):
        if not is_prime(number, primes):
            if not is_sum_of_prime_and_twice_a_square(number, primes, double_squares):
                return number
    raise RuntimeError("No solution found")

File: src/P046/p046.go
package problem046

import (
	"strconv"

	"github.com/sunilbpandey/project-euler/src/utils/go/math/primeutils"
)

func isSumOfPrimeAndTwiceASquare(n int, primes []int, doubleSquares []int) bool {
	for _, dsq := range doubleSquares {
		if dsq > n {
			break
		}

		if primeutils.IsPrime(n-dsq, primes) {
			return true
		}
	}
	return false
}

func Solve() string {
	primes := primeutils.GeneratePrimes(1000)
	doubleSquares := make([]int, 1000)
	for i := 0; i < 1000; i++ {
		doubleSquares[i] = 2 * i * i
	}

	answer := 0
	for n := 9; n < 1000000; n += 2 {
		if !primeutils.IsPrime(n, primes) && !isSumOfPrimeAndTwiceASquare(n, primes, doubleSquares) {
			answer = n
			break
		}
	}
	return strconv.Itoa(answer)
}
